### PR TITLE
feat(ingest/salesforce): helpful error messages on failure

### PR DESCRIPTION
**Pull request by mayurinehate**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/salesforce.py [hunk 1]
```diff
@@ -287,7 +287,13 @@ def __init__(self, config: SalesforceConfig, ctx: PipelineContext) -> None:
 
         except Exception as e:
             logger.error(e)
-            raise ConfigurationError("Salesforce login failed") from e
+            if "API_CURRENTLY_DISABLED" in str(e):
```
Comment (skrydal) [inline metadata-ingestion/src/datahub/ingestion/source/salesforce.py:290]: This string will be present **only if** the `Exception` is actually:
`simple_salesforce.exceptions.SalesforceAuthenticationFailed` maybe we can catch is explicitly?

File: metadata-ingestion/src/datahub/ingestion/source/salesforce.py [hunk 2]
```diff
@@ -314,10 +320,19 @@ def __init__(self, config: SalesforceConfig, ctx: PipelineContext) -> None:
...
+            if "sObject type 'EntityDefinition' is not supported." in str(e):
+                # https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/tooling_api_objects_entitydefinition.htm
+                raise ConfigurationError(
+                    "Salesforce EntityDefinition query failed. "
+                    "Please verify if user has 'View Setup and Configuration' permission."
+                ) from e
+            raise e
+        else:
+            for sObject in sObjects:
+                yield from self.get_salesforce_object_workunits(sObject)
```
Comment (skrydal) [inline metadata-ingestion/src/datahub/ingestion/source/salesforce.py:335]: Awesome 👍

**Review by skrydal — CHANGES_REQUESTED** [review]
Looks very good, left some comments about clarification of error messages trigger.

File: metadata-ingestion/src/datahub/ingestion/source/salesforce.py [hunk 3]
```diff
@@ -287,7 +287,13 @@ def __init__(self, config: SalesforceConfig, ctx: PipelineContext) -> None:
 
         except Exception as e:
             logger.error(e)
-            raise ConfigurationError("Salesforce login failed") from e
+            if "API_CURRENTLY_DISABLED" in str(e):
+                # https://help.salesforce.com/s/articleView?id=001473830&type=1
+                error = "Salesforce login failed. Please make sure user has API Enabled Access."
+            else:
+                error = "Salesforce login failed. Please verify your credentials."
+                "Please set `is_sandbox: True` in recipe if this is sandbox account."
```
[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/salesforce.py:295]: I'm not quite sure if salesforce sandbox url follows any specific structure. Do you have any resource that confirms that instance url can be used to identify whether this is sandbox account ?

[the remark this answers is not in the document]
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/salesforce.py:295]: Noticed that its general pattern to use "sandbox" in name, so in favour of best practice, will change this handling to check whether "sandbox" is present in url, before showing this message.

**Review by skrydal — APPROVED** [review]
LGTM

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**